source/matrix_free_utils.cc

```diff
@@ -0,0 +1,249 @@
+
+
+template <int dim, typename number, int n_components>
+CouplingOperator<dim, number, n_components>::CouplingOperator(
+  const Inclusions<dim>           &inclusions_,
+  const DoFHandler<dim>           &dof_handler_,
+  const AffineConstraints<number> &constraints_,
+  const MappingQ<dim>             &mapping_,
+  const FiniteElement<dim>        &fe_)
+  : rpe{{1e-9, true, inclusions_.rtree_extraction_level, {}}}
+  , n_coefficients{inclusions_.n_coefficients}
+{
+  [[maybe_unused]] auto parallel_triangulation =
+    dynamic_cast<const parallel::distributed::Triangulation<dim, dim> *>(
+      &dof_handler_.get_triangulation());
+  Assert((parallel_triangulation != nullptr),
+         ExcMessage("Only p::d::T triangulations are supported."));
+
+  mapping     = &mapping_;
+  fe          = &fe_;
+  constraints = &constraints_;
+  dof_handler = &dof_handler_;
+  inclusions  = &inclusions_;
+
+  //  Get all locally owned support points from the inclusions
+  std::vector<Point<dim>> locations(
+    inclusions->inclusions_as_particles.n_locally_owned_particles());
+  const_cast<Particles::ParticleHandler<dim> &>(
+    inclusions->inclusions_as_particles)
+    .get_particle_positions(locations);
+
+  rpe.reinit(locations, dof_handler->get_triangulation(), *mapping);
+
+  // We should make sure all points have been found. Do this only in debug mode
+  Assert(rpe.all_points_found(),
+         ExcInternalError("Not all points have been"
+                          "found during the geometric search procedure."));
+}
+
+
+
+template <int dim, typename number, int n_components>
+void
+CouplingOperator<dim, number, n_components>::initialize_dof_vector(
+  VectorType &vec) const
+{
+  (void)vec;
+}
+
+
+
+template <int dim, typename number, int n_components>
+void
+CouplingOperator<dim, number, n_components>::vmult(VectorType       &dst,
+                                                   const VectorType &src) const
+{
+  Assert(rpe.is_ready(),
+         ExcInternalError("Remote evaluator has not been initialized."));
+
+  src.update_ghost_values();
+  dst = 0.;
+
+  std::vector<number> integration_values;
+  const unsigned int  n_dofs_per_cell = fe->n_dofs_per_cell();
+
+  // collect
+  std::vector<types::global_dof_index> inclusion_dof_indices(
+    inclusions->n_coefficients);
+
+  auto particle = inclusions->inclusions_as_particles.begin();
+  while (particle != inclusions->inclusions_as_particles.end())
+    {
+      const auto id                   = particle->get_id();
+      inclusion_dof_indices           = inclusions->get_dof_indices(id);
+      const auto &inclusion_fe_values = inclusions->get_fe_values(id);
+
+      double value_per_particle = 0.;
+      for (unsigned int j = 0; j < inclusions->n_coefficients; ++j)
+        {
+          value_per_particle +=
+            inclusion_fe_values[j] * src(inclusion_dof_indices[j]);
+        }
+      integration_values.push_back(value_per_particle);
+      ++particle;
+    }
```

Comment:
This part is still using particles, and not rpe. Is this expected?

Reply:
Yes, rpe is responsible for testing and integrating over the points found during the search procedure. Before that, we need to collect the values to be tested and integrated over.